content/blog/2020/10/2020-10-31-spring-xstream.adoc

```diff
@@ -0,0 +1,67 @@
+---
+
+Cleaning up technical debt is a perennial topic among Jenkins core developers,
+and one of the most visible issues is the use of obsolete and/or forked third-party libraries.
+In a world where Dependabot is offering updates to libraries released just hours before,
+it is unpleasant to be working with dependencies that are many years old.
+Since large organizations in particular are unhappy to install software using obsolete or nonstandard versions,
+my employer (CloudBees) gave its blessing for me to spend some time cleaning up egregious cases.
+
+The toughest nut to crack was the Acegi Security library used for authentication,
+which has long since been replaced by Spring Security
+(and Jenkins was also bundling a long-outdated version of some Spring Framework dependencies).
+jep:227[] tracks the complicated task of updating to Spring Security
+without breaking the numerous plugins that interact with authentication,
+especially those offering a *Security Realm*.
+
+Another longstanding problem was the XStream library which Jenkins uses to read and write XML configuration files.
+This had been forked long ago by what was then the Hudson project and a few fixes applied.
+Unfortunately, some of those fixes were rejected upstream as invalid (representing unsupported usage patterns),
+and the fork fell behind the upstream version.
+// TODO jep:nnn[] after https://github.com/jenkinsci/jep/pull/309 is merged
+A link:https://github.com/jglick/jep/blob/xstream/jep/0000/README.adoc[second JEP]
+describes the impact of switching to the upstream library in a more standard usage mode,
+including fixes to a smaller number of plugins which would otherwise be incompatible.
+
+Now that Jenkins weekly releases include both updates,
+// TODO specify version after https://github.com/jenkinsci/jenkins/pull/4848 and https://github.com/jenkinsci/jenkins/pull/4944 are merged
+it is important for both Jenkins administrators and plugin maintainers to check for actual or potential incompatibilities.
+There are two tables listing the impact of these changes on plugins:
+
+* link:https://github.com/jenkinsci/jep/blob/master/jep/227/compatibility.adoc[Spring Security compatibility]
+// TODO switch link after https://github.com/jenkinsci/jep/pull/309 is merged
+* link:https://github.com/jglick/jep/blob/xstream/jep/0000/compatibility.adoc[XStream compatibility]
+
+If you *use Jenkins* then it is a good idea before upgrading to take a look at these tables
+to see if you are running any plugins considered _incompatible_.
+If so, try not to rely on that plugin, or find out if there is an active maintainer who could help.
+For entries marked _unknown_, it would be appreciated if you could do a sanity check after upgrading
+and offer a pull request to the table page (click *Edit this file*) with a more informative status.
+If you find a regression in a plugin, please file a bug report in Jira and link to it from the table.
```

Comment:
💡 Do we want to propose a label to reduce the likelihood of tickets being not seen?
It was done for JEP-200 with the label `JEP-200`.

Reply:
Yeah we can do that, especially once we have a number registered for the XStream JEP.